src/v/raft/mutex_buffer.h

```diff
@@ -0,0 +1,130 @@
+        { f(std::move(req)) } -> ss::future<Response>;
+    })
+    // clang-format on
+    void start(Func&& f);
+
+    ss::future<> stop() {
+        _enequeued.broken();
+        return _gate.close();
+    }
+
+private:
+    using request_t = std::vector<Request>;
+    using response_t = std::vector<ss::promise<Response>>;
+    template<typename Func>
+    ss::future<> flush(Func&& f);
+    template<typename Func>
+    ss::future<> do_flush(request_t, response_t, Func&& f);
+
+    request_t _requests;
+    response_t _responsens;
+    ss::condition_variable _enequeued;
+    ss::gate _gate;
+    mutex& _mutex;
+    const size_t _max_buffered;
+};
+
+template<typename Request, typename Response>
+template<typename Func>
+void mutex_buffer<Request, Response>::start(Func&& f) {
+    (void)ss::with_gate(_gate, [this, f = std::forward<Func>(f)]() mutable {
+        return ss::do_until(
+          [this] { return _gate.is_closed(); },
+          [this, f = std::forward<Func>(f)]() mutable {
+              return _enequeued.wait([this] { return !_requests.empty(); })
+                .then([this, f = std::forward<Func>(f)]() mutable {
+                    return flush(std::forward<Func>(f));
+                })
+                .handle_exception_type(
+                  [](const ss::broken_condition_variable&) {
+                      // ignore exception, we are about to stop
+                  });
+          });
+    });
+}
+template<typename Request, typename Response>
+template<typename Func>
+ss::future<> mutex_buffer<Request, Response>::flush(Func&& f) {
+    auto requests = std::exchange(_requests, {});
+    auto response_promises = std::exchange(_responsens, {});
+
+    return _mutex.with(
+      [this,
+       f = std::forward<Func>(f),
+       requests = std::move(requests),
+       response_promises = std::move(response_promises)]() mutable {
+          return do_flush(
+            std::move(requests),
+            std::move(response_promises),
+            std::forward<Func>(f));
+      });
+}
+
+template<typename Request, typename Response>
+template<typename Func>
+ss::future<> mutex_buffer<Request, Response>::do_flush(
+  request_t requests, response_t response_promises, Func&& f) {
+    auto resp = ss::make_lw_shared(std::move(response_promises));
+    return ss::do_with(
+             std::move(requests),
+             resp->begin(),
+             [f = std::forward<Func>(f)](
+               request_t& requests, typename response_t::iterator& it) mutable {
+                 return ss::do_for_each(
+                   requests,
+                   [&it, f = std::forward<Func>(f)](Request& req) mutable {
+                       return f(std::move(req))
+                         .then([&it](Response r) {
+                             (it++)->set_value(std::move(r));
+                         })
+                         .handle_exception([&it](const std::exception_ptr& e) {
+                             (it++)->set_exception(e);
```

Comment:
I think this is a bug. you want `(it--)->set_exception()`

Why not capture a copy of the iterator? 

auto current = it;
it++;
and use `current` in both lambdas. it's always correct and easier to read.

Reply:
ah nvm.